File: BOX.py
import numpy as np
import Objects
from numpy.typing import NDArray
from typing import List as List, Tuple

def generate_unique_triplets(N: int, L: int) -> np.ndarray:
    """
    Generate N unique triplets of integers (x, y, z) where 0 <= x, y, z <= L.

    Args:
        N (int): Number of unique triplets to generate.
        L (int): Maximum value for each element in the triplet (0 to L).

    Returns:
        numpy.ndarray: Array of shape (N, 3) containing N unique triplets.
    """
    # Total possible unique triplets
    total_triplets = (L + 1) ** 3
    
    # Check if N is larger than the number of possible unique triplets
    if N > total_triplets:
        raise ValueError("N is larger than the total number of unique triplets available.")
    
    # Generate N unique indices within the range of possible triplets
    selected_indices = np.random.choice(total_triplets, size=N, replace=False)
    
    # Convert the indices back to triplets (x, y, z)
    z = selected_indices % (L + 1)
    y = (selected_indices // (L + 1)) % (L + 1)
    x = (selected_indices // ((L + 1) * (L + 1))) % (L + 1)
    
    # Stack them into an N x 3 array
    return np.stack((x, y, z), axis=-1)

def to_single_index(x: int, y: int, z: int, L: int) -> int:
    """Convert 3D indices (x, y, z) to a single index."""
    return x * L * L + y * L + z

def to_xyz(index: int, L: int) -> Tuple[int, int, int]:
    """Convert a single index back to 3D indices (x, y, z)."""
    z = index % L
    y = (index // L) % L
    x = index // (L * L)
    return x, y, z

class box:
    def __init__(self, 
                 size: int,
                 nobjects:int,
                 Interactions: List[List[float]]) -> None:
        self.size = size
        self.lattice = np.zeros((self.size, self.size, self.size), dtype=Objects.Object)
        for x in range(self.size):
            for y in range(self.size):
                for z in range(self.size):
                    self.lattice[x,y,z] = Objects.Empty((x,y,z))
        
        self.objects = np.empty((nobjects), dtype=Objects.Object)

        self.E = Interactions # symetric matrix of unspecific interactions

    def set_lattice(self,site:Tuple[int,int,int],object:Objects)->None:
        self.lattice[site]=object
        object.position = site

    def get_lattice(self,site:Tuple[int,int,int])->None:
        return self.lattice[site]
        

    def compute_local_energy(self, xyz:Tuple) -> float:
        """ Computes the local energy for a given lattice site. """
        if self.lattice[xyz].isempty():
            return 0.0
        neighbors = self.get_neighbors(xyz)
        local_energy = 0.0


        neigh_count = 0.0 # for later
        for nxyz in neighbors:            
            local_energy -= self.E[self.lattice[xyz].Index()][self.lattice[nxyz].Index()]

        return local_energy
    def total_energy(self) -> float:
        """ Computes the total energy of the system. 
            The function is really not efficient, but only used once at the beginning.
        """
        energy = 0.0
        for x in range(self.size):
            for y in range(self.size):
                for z in range(self.size):
                    if not self.lattice[x,y,z].isempty():
                        xyz=(x,y,z)
                        energy += self.compute_local_energy(xyz)
        return energy / 2  # To correct for double counting of pairs


    def get_neighbors(self, xyz:Tuple[int,int,int]) -> List[np.ndarray]:
        """ Returns the coordinates of the 26 nearest neighbors of a given lattice site (x, y, z). """
        neighbors = []
        for dx in [-1, 0, 1]:
            for dy in [-1, 0, 1]:
                for dz in [-1, 0, 1]:
                    if dx == 0 and dy == 0 and dz == 0:
                        continue
                    neighbors.append(((xyz[0] + dx) % self.size, 
                                                (xyz[1] + dy) % self.size, 
                                                (xyz[2] + dz) % self.size))
        return neighbors

    def has_free_neighbor(self, xyz:Tuple[int,int,int])-> bool:
        """
        Check if a site has a surounding free site
        """
        neighbors = self.get_neighbors(xyz)
        for nxyz in neighbors:
            if self.lattice[nxyz].isempty():
                return True
        return False

        
    def build_clusters(self) -> Tuple[NDArray[np.int_], NDArray[np.int_]]:
        """
        Computes the distribution of cluster sizes in the lattice.
        A cluster is defined as a set of connected indices where lattice[index] = True.
        The connected condition is defined by the get_neighbors function.

        Returns:
            Tuple[NDArray[np.int_], NDArray[np.int_]]: A tuple containing an array of cluster indices and an array of cluster start indices.
        """
        visited = np.zeros(self.size ** 3, dtype=np.bool_)  # Flattened 3D lattice
        cluster_indices = np.empty(self.size ** 3, dtype=np.int32)  # Allocate maximum size, we'll trim later
        cluster_starts = np.empty(self.size ** 3, dtype=np.int32)  # Same here
        cluster_count = 0
        current_index = 0

        def flood_fill(start_index: int) -> int:
            """Performs flood fill from the start index to find the entire cluster."""
            nonlocal current_index,cluster_indices,visited,cluster_starts,cluster_count
            stack = list([start_index])
            cluster_start = current_index

            while len(stack) > 0:
                current_idx = stack.pop()
                if not visited[current_idx]:
                    visited[current_idx] = True
                    cluster_indices[current_index] = current_idx
                    current_index += 1
                    x, y, z = to_xyz(current_idx, self.size)
                    for neighbor in self.get_neighbors((x, y, z)):
                        nx, ny, nz = neighbor[0], neighbor[1], neighbor[2]
                        neighbor_idx = to_single_index(nx, ny, nz, self.size)
                        if self.lattice[nx, ny, nz] and not visited[neighbor_idx]:
                            stack.append(neighbor_idx)
            if current_index > cluster_start:
                cluster_starts[cluster_count] = cluster_start
                cluster_count += 1

        # Iterate over all lattice sites
        for x in range(self.size):
            for y in range(self.size):
                for z in range(self.size):
                    idx = to_single_index(x, y, z, self.size)
                    if self.lattice[x, y, z] and not visited[idx]:
                        # Start a new cluster
                        flood_fill(idx)

        return cluster_indices[:current_index], cluster_starts[:cluster_count]
            

    def cluster_size(self) -> NDArray[np.int32]:
        cluster_indices, cluster_starts = self.build_clusters()
        cluster_starts = np.append(cluster_starts, len(cluster_indices))
        sizes = np.diff(cluster_starts)
        return sizes
    def compute_av_Nneigh(self) -> int:
        av_Nneigh= 0
        for x in range(self.size):
            for y in range(self.size):
                for z in range(self.size):
                    if self.lattice[x,y,z]:
                        counter = 0
                        for neighbor in self.get_neighbors((x, y, z)):
                            nx, ny, nz = neighbor[0], neighbor[1], neighbor[2]
                            if self.lattice[nx,ny,nz]:
                                counter+=1
                        av_Nneigh+=counter/self.true_sites.shape[0]
        return av_Nneigh

File: BOX.cpp
// BOX.cpp
#include "BOX.h"
#include "Objects.h"
#include <stdexcept>
#include <random>
#include <algorithm>
#include <iostream>
#include <unordered_set>
#include <cmath>

int to_single_index(int x, int y, int z, int L) {
    return x * L * L + y * L + z;
}

std::tuple<int, int, int> to_xyz(int index, int L) {
    int z = index % L;
    int y = (index / L) % L;
    int x = index / (L * L);
    return std::make_tuple(x, y, z);
}

std::vector<std::array<int, 3>> generate_unique_triplets(int N, int L) {
    int total_triplets = (L + 1) * (L + 1) * (L + 1);

    if (N > total_triplets) {
        throw std::invalid_argument("N is larger than the total number of unique triplets available.");
    }

    std::vector<int> indices(total_triplets);
    for (int i = 0; i < total_triplets; ++i) {
        indices[i] = i;
    }

    static std::mt19937 rng(std::random_device{}());
    std::shuffle(indices.begin(), indices.end(), rng);

    std::vector<std::array<int, 3>> triplets(N);
    for (int i = 0; i < N; ++i) {
        int selected_index = indices[i];

        int z = selected_index % (L + 1);
        int y = (selected_index / (L + 1)) % (L + 1);
        int x = (selected_index / ((L + 1) * (L + 1))) % (L + 1);

        triplets[i] = { x, y, z };
    }
    return triplets;
}

BOX::BOX(int size_, int nobjects, const std::vector<std::vector<float>>& Interactions)
    : size(size_), E(Interactions){
    int total_sites = size * size * size;
    lattice.resize(total_sites);

    // Initialize lattice with Empty objects
    for (int x = 0; x < size; ++x) {
        for (int y = 0; y < size; ++y) {
            for (int z = 0; z < size; ++z) {
                int idx = to_single_index(x, y, z, size);
                lattice[idx] = new Empty(std::make_tuple(x, y, z));
            }
        }
    }

    // Initialize objects array
    objects.resize(nobjects, nullptr);
}

BOX::~BOX() {
    // Delete all unique Objects in lattice
    std::unordered_set<Object*> unique_objects;
    for (auto obj : lattice) {
        unique_objects.insert(obj);
    }
    for (auto obj : unique_objects) {
        delete obj;
    }
}

void BOX::create_new_DHH1(const std::tuple<int,int,int>& site, int object_idx){
    DHH1* new_dhh1 = new DHH1(site);
    int idx = to_single_index(std::get<0>(site), std::get<1>(site), std::get<2>(site), size);
    delete lattice[idx];
    lattice[idx] = new_dhh1;
    objects[object_idx] = new_dhh1;
    new_dhh1->setPosition(site);
}

void BOX::swap(const std::tuple<int, int, int>& site1, const std::tuple<int, int, int>& site2) {
    int idx1 = to_single_index(std::get<0>(site1), std::get<1>(site1), std::get<2>(site1), size);
    int idx2 = to_single_index(std::get<0>(site2), std::get<1>(site2), std::get<2>(site2), size);

    std::swap(lattice[idx1], lattice[idx2]);

    // Update the positions of the objects
    lattice[idx1]->setPosition(site1);
    lattice[idx2]->setPosition(site2);
}

Object* BOX::get_lattice(const std::tuple<int, int, int>& site) const {
    int x = std::get<0>(site);
    int y = std::get<1>(site);
    int z = std::get<2>(site);
    int idx = to_single_index(x, y, z, size);
    return lattice[idx];
}

float BOX::compute_local_energy(const std::tuple<int, int, int>& xyz) const {
    Object* obj = get_lattice(xyz);
    if (obj->isempty()) {
        return 0.0f;
    }
    auto neighbors = get_neighbors(xyz);
    float local_energy = 0.0f;

    for (const auto& nxyz : neighbors) {
        Object* neighbor_obj = get_lattice(nxyz);
        local_energy -= E[obj->Index()][neighbor_obj->Index()];
    }

    return local_energy;
}

float BOX::total_energy() const {
    float energy = 0.0f;
    for (int x = 0; x < size; ++x) {
        for (int y = 0; y < size; ++y) {
            for (int z = 0; z < size; ++z) {
                Object* obj = get_lattice(std::make_tuple(x, y, z));
                if (!obj->isempty()) {
                    energy += compute_local_energy(std::make_tuple(x, y, z));
                }
            }
        }
    }
    return energy / 2.0f;  // Correct for double counting
}

std::vector<std::tuple<int, int, int>> BOX::get_neighbors(const std::tuple<int, int, int>& xyz) const {
    int x = std::get<0>(xyz);
    int y = std::get<1>(xyz);
    int z = std::get<2>(xyz);

    std::vector<std::tuple<int, int, int>> neighbors;
    for (int dx = -1; dx <=1; ++dx) {
        for (int dy = -1; dy <=1; ++dy) {
            for (int dz = -1; dz <=1; ++dz) {
                if (dx == 0 && dy == 0 && dz == 0) {
                    continue;
                }
                int nx = (x + dx + size) % size;
                int ny = (y + dy + size) % size;
                int nz = (z + dz + size) % size;
                neighbors.push_back(std::make_tuple(nx, ny, nz));
            }
        }
    }
    return neighbors;
}

bool BOX::has_free_neighbor(const std::tuple<int, int, int>& xyz) const {
    auto neighbors = get_neighbors(xyz);
    for (const auto& nxyz : neighbors) {
        Object* neighbor_obj = get_lattice(nxyz);
        if (neighbor_obj->isempty()) {
            return true;
        }
    }
    return false;
}

// Placeholder implementations for build_clusters, cluster_size, and compute_av_Nneigh

std::tuple<std::vector<int>, std::vector<int>> BOX::build_clusters() {
    // Implement cluster building algorithm
    return std::make_tuple(std::vector<int>(), std::vector<int>());
}

std::vector<int> BOX::cluster_size() {
    auto [cluster_indices, cluster_starts] = build_clusters();
    std::vector<int> sizes;
    // Compute sizes based on cluster_starts and cluster_indices
    return sizes;
}

int BOX::compute_av_Nneigh() const {
    int av_Nneigh = 0;
    int true_sites = 0;
    for (int x = 0; x < size; ++x) {
        for (int y = 0; y < size; ++y) {
            for (int z = 0; z < size; ++z) {
                Object* obj = get_lattice(std::make_tuple(x, y, z));
                if (!obj->isempty()) {
                    int counter = 0;
                    auto neighbors = get_neighbors(std::make_tuple(x, y, z));
                    for (const auto& neighbor : neighbors) {
                        Object* neighbor_obj = get_lattice(neighbor);
                        if (!neighbor_obj->isempty()) {
                            ++counter;
                        }
                    }
                    av_Nneigh += counter;
                    ++true_sites;
                }
            }
        }
    }
    return true_sites > 0 ? av_Nneigh / true_sites : 0;
}
